### PR TITLE
Support `for`-`in`-`if` expressions

### PR DESCRIPTION
**Pull request by radeksimko**
Closes https://github.com/hashicorp/terraform-ls/issues/527

--- 

## Follow-up Issues

 - [ ] https://github.com/hashicorp/terraform-ls/issues/1607
 - [ ] https://github.com/hashicorp/terraform-ls/issues/1608
 - [ ] https://github.com/hashicorp/terraform-ls/issues/1609
 - [ ] https://github.com/hashicorp/hcl-lang/issues/370

### CODE REVIEW

File: decoder/expr_any_completion_test.go
Comment (dbanck) [inline decoder/expr_any_completion_test.go:4083]: I found this confusing at first, because I was expecting mainly `v` to be used here. But after giving it some thought, it makes sense to have the full completion list here.

File: decoder/expr_any_for.go
Comment (dbanck) [inline decoder/expr_any_for.go:21]: Can we recover the `.` when triggering completion with a prefix?

CleanShot 2024-02-06 at 10 24 57@2x

**Review by dbanck — APPROVED** [review]
Nice work!

Just a question/suggestion that could also be addressed in a follow-up issue.

File: decoder/expr_any_completion_test.go
Comment (radeksimko) [reply decoder/expr_any_completion_test.go:4083]: The `v` as in the value reference support is something now tracked under https://github.com/hashicorp/terraform-ls/issues/1608

File: decoder/expr_any_for.go
Comment (radeksimko) [reply decoder/expr_any_for.go:21]: That would be a little more difficult here unfortunately as we'd have to be recovering pretty much everything up to the `[` _at least_. I didn't even think of the consequences of the incomplete `ForExpr` wrapped into another expression like function here but I expect it to be similarly difficult.

Either way we now track it under https://github.com/hashicorp/hcl-lang/issues/370

**Comment by radeksimko** [comment]
@dbanck I no longer have push/merge permissions here, so I'll have to leave it up to you to merge 😅

**Comment by lukaabicic** [comment]
Please fix variable declaration not found error in [2.29.4] version (2024-02-13). 
The error occurs when the resources are created with a for_each argument whose values are variables of type list(object()).
The error isn't present in the previous versions.

**Pull request merged**
